sanic/server/protocols/websocket_protocol.py

```diff
@@ -0,0 +1,113 @@
+        self.websocket_write_limit = websocket_write_limit
+        self.websocket_ping_interval = websocket_ping_interval
+        self.websocket_ping_timeout = websocket_ping_timeout
+
+    def connection_lost(self, exc):
+        if self.websocket is not None:
+            self.websocket.connection_lost(exc)
+        super().connection_lost(exc)
+
+    def data_received(self, data):
+        if self.websocket is not None:
+            self.websocket.data_received(data)
+        else:
+            # Pass it to HttpProtocol handler first
+            # That will (hopefully) upgrade it to a websocket.
+            super().data_received(data)
+
+    def eof_received(self) -> Optional[bool]:
+        if self.websocket is not None:
+            return self.websocket.eof_received()
+        else:
+            return False
+
+    def close(self, timeout: Optional[float] = None):
+        # Called by HttpProtocol at the end of connection_task
+        # If we've upgraded to websocket, we do our own closing
+        if self.websocket is not None:
+            self.websocket.fail_connection(1001)
+        else:
+            super().close()
+
+    def close_if_idle(self):
+        # Called by Sanic Server when shutting down
+        # If we've upgraded to websocket, shut it down
+        if self.websocket is not None:
+            if self.websocket.connection.state in (CLOSING, CLOSED):
+                return True
+            else:
+                return self.websocket.fail_connection(1001)
+        else:
+            return super().close_if_idle()
+
+    async def websocket_handshake(self, request, subprotocols=Optional[Sequence[str]]):
+        # let the websockets package do the handshake with the client
+        headers = {"Upgrade": "websocket", "Connection": "Upgrade"}
+        try:
+            if subprotocols is not None:
+                # subprotocols can be a set or frozenset, but ServerConnection needs a list
+                subprotocols = list(subprotocols)
+            ws_conn = ServerConnection(max_size=self.websocket_max_size, subprotocols=subprotocols,
+                                       state=OPEN, logger=error_logger)
+            resp: "http11.Response" = ws_conn.accept(request)
+        except Exception as exc:
+            msg = (
+                    "Failed to open a WebSocket connection.\n"
+                    "See server log for more information.\n"
+                )
+            raise SanicException(msg, status_code=500)
+        if 100 <= resp.status_code <= 299:
+            rbytes = b"".join([b"HTTP/1.1 ", b'%d' % resp.status_code, b" ", resp.reason_phrase.encode("utf-8"), b"\r\n"])
+            for k, v in resp.headers.items():
+                rbytes += k.encode("utf-8") + b": " + v.encode("utf-8") + b"\r\n"
+            if resp.body:
+                rbytes += b"\r\n" + resp.body + b"\r\n"
+            rbytes += b"\r\n"
+            await super().send(rbytes)
```

Comment:
Now that I am looking at this whole block, I would use regular strings up until this point and just encode once after it is constructed.

Reply:
Yes, much faster to encode once all is finished. Also I prefer not saying "utf-8" in the argument because that goes without saying (UTF-8 is the default).

It is also faster to combine different stages and to use f-strings, so lines 103-105 would be

```python
if resp.body is not None:
    rbytes += f"\r\n{resp.body}\r\n\r\n"
else:
    rbytes += "\r\n"
```

Reply:
And also faster to

```python
rbytes += "".join(f"{k}: {v}/r/n" for k, v in resp.headers.items())
```

I think this code already exists for HTTP response formatting in Sanic, perhaps reuse the existing function?